music21/osmd/osmd.py

```diff
@@ -0,0 +1,233 @@
+
+class OpenSheetMusicDisplayException(exceptions21.Music21Exception):
+    pass
+
+
+# Determine if we are running inside an IPython notebook
+# if so will render inline - if not we open a new browser window
+try:
+    loader = importlib.util.find_spec('IPython.core.display')
+except ImportError:
+    loader = None
+hasInstalledIPython = loader is not None
+
+
+def getExtendedModules():
+    # pylint: disable=import-outside-toplevel
+    if hasInstalledIPython:
+        from IPython.core.display import display, HTML, Javascript
+    else:
+        def display():
+            raise OpenSheetMusicDisplayException(
+                'OpenSheetMusicDisplay requires IPython to be installed')
+
+        HTML = Javascript = display
+    return display, HTML, Javascript
+
+
+class ConverterOpenSheetMusicDisplay(SubConverter):
+    '''
+    Converter for displaying sheet music inline
+    in an IPython notebook.
+    Uses: https://opensheetmusicdisplay.org/ for rendering
+    '''
+    registerFormats = ('osmd',)
+    registerShowFormats = ('osmd',)
+    # when updating the script tag osmd will only reload in a notebook if you: Kernel -> restart and
+    # clear output, then save the notebook and refresh the page.
+    # Otherwise the script will stay on the page and not reload.
+    script_url = ("https://github.com/opensheetmusicdisplay"
+                  + "/opensheetmusicdisplay/releases/download/0.9.2/opensheetmusicdisplay.min.js")
+
+    def __init__(self):
+        super().__init__()
+        self.display, self.HTML, self.Javascript = getExtendedModules()
+
+    def show(self, obj, fmt, *,
+             fixPartName=True, offline=False, divId=None,
+             **keywords):
+        '''
+        Displays the score object in a notebook using the OpenSheetMusicDisplay.js library.
+
+        >>> import music21
+        >>> s = music21.converter.parse("tinyNotation: 3/4 E4 r f# g=lastG trip{b-8 a g} c4~ c")
+        >>> fig_id1 = s.show('osmd')
+
+        To update a previously displayed score use the returned divId from before:
+
+        >>> fig_id2 = s.show('osmd', divId=fig_id1)
+        >>> assert fig_id1 == fig_id2
+        '''
+        in_ipython = runningUnderIPython()
+        score = obj
+        if fixPartName:
+            self.addDefaultPartName(score)
+
+        if divId is None:
+            # create unique reference to output div in case we wish to update it
+            divId = self.getUniqueDivId()
+
+        # convert score to xml string
+        gex = m21ToXml.GeneralObjectExporter(score)
+        bytesOut = gex.parse()
+        xml = bytesOut.decode('utf-8')
+
+        # generate script to be run on page
+        script = self.musicXMLToScript(xml, divId, offline=offline)
+        if in_ipython:
+            self.display(self.HTML(f'<div id="{divId}"></div>'))
+            self.display(self.Javascript(script))
+        else:
+
+            tmp = tempfile.NamedTemporaryFile(delete=False)
+            tmp_path = tmp.name + '.html'
```

Comment:
Still think it would be a good idea to have the temp files all in one place / observe the user's scratch directory.

```suggestion
            tmp = self.getTemporaryFile()
```

Problem is, we can't do that until we register the subconverter, I'll reply to the earlier comment about this.

```
>>> osmdc.getTemporaryFile(subformats=['html'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/music21/converter/subConverters.py", line 242, in getTemporaryFile
    ext = self.getExtensionForSubformats(subformats)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/music21/converter/subConverters.py", line 219, in getExtensionForSubformats
    raise SubConverterException(
music21.converter.subConverters.SubConverterException: This subconverter cannot show or write: no output extensions are registered for it
```